src/dag/read.rs

```diff
@@ -0,0 +1,54 @@
+use super::chunk::Chunk;
+use super::key::Key;
+use super::{Error, Result};
+use crate::kv;
+
+#[allow(dead_code)]
+pub struct Read<'a> {
+    kvr: Box<dyn kv::Read + 'a>,
+}
+
+#[allow(dead_code)]
+impl<'a> Read<'_> {
+    pub fn new(kvr: Box<dyn kv::Read + 'a>) -> Read {
+        Read{kvr}
+    }
+
+    pub async fn has_chunk(&mut self, hash: &str) -> Result<bool> {
+        has_chunk(self.kvr.as_ref(), hash).await
+    }
+
+    pub async fn get_chunk(&mut self, hash: &str) -> Result<Option<Chunk>> {
+        get_chunk(self.kvr.as_ref(), hash).await
+    }
+
+    pub async fn get_head(&mut self, name: &str) -> Result<Option<String>> {
+        get_head(self.kvr.as_ref(), name).await
+    }
+}
+
+pub async fn has_chunk(kvr: &dyn kv::Read, hash: &str) -> Result<bool> {
```

Comment:
These have to be factored out as bare functions because Rust:

https://users.rust-lang.org/t/casting-traitobject-to-super-trait/33524

Reply:
As with so many things Rust, I'm sure there's a way to do this, but it ends up being easier to just not use advanced features.